crates/snapbox/src/data.rs

```diff
@@ -24,6 +30,47 @@ impl Default for DataFormat {
     }
 }
 
+#[cfg(feature = "structured-data")]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub enum Value {
+    Null,
+    Bool(bool),
+    Number(serde_json::Number),
+    String(String),
+    Array(Vec<Value>),
+    // IndexMap is used as it keeps insertion order
+    Object(indexmap::IndexMap<String, Value>),
+}
+
+#[cfg(feature = "structured-data")]
+impl Value {
+    fn normalize_paths(self) -> Self {
+        match self {
+            Value::Null => Value::Null,
+            Value::Bool(b) => Value::Bool(b),
+            Value::Number(num) => Value::Number(num),
+            Value::String(str) => {
+                let str = crate::utils::normalize_paths(&str);
+                Value::String(str)
+            }
+            Value::Array(arr) => {
+                let arr = arr
+                    .into_iter()
+                    .map(|value| value.normalize_paths())
+                    .collect();
+                Value::Array(arr)
+            }
+            Value::Object(obj) => {
+                let obj = indexmap::IndexMap::from_iter(
+                    obj.into_iter()
+                        .map(|(str, value)| (str, value.normalize_paths())),
+                );
+                Value::Object(obj)
+            }
```

Comment:
Can we do this without destroying and recreating the wrapping data structures?

Reply:
I think it is possible to do so. It might also be a good idea to make it more generic across normalization operations by taking in what operation needs done.